### ISSUE TITLE
Refactor: extend model training API

### ISSUE THREAD

**Issue by sgreenbury** [alan-turing-institute/autoemulate#475]
Currently skorch is used to provide functionality such as callbacks like early stopping and learning rate scheduling for NN models to facilitate regularization and optimize training.

With the refactored version we could consider the following:
- Adding custom AutoEmulate API for this feature
- Adding a dep (e.g. [lightning](https://lightning.ai/docs/pytorch/stable/common/trainer.html))

Also relates to #329.

**Comment by radka-j** [comment]
We should definitely use PyTorch Lightning for this.

**Comment by sgreenbury** [comment]
Looking at the [lightning docs](https://lightning.ai/docs/pytorch/stable/common/lightning_module.html), adding some thoughts on how we might incorporate:
- Since a `LightningModule` is a subclass of `nn.module`, we might consider replacing `nn.Module` with `LightningModule` in `PyTorchBackend` for the subclass.
- It would be great if we could also use `LightningModule` with the gpytorch too. This might be slightly more complicated as the gpytorch base classes (e.g. `gpytorch.models.ExactGP`) already inherit from `nn.module` in possibly incompatible way (we previously did not inherit from `PyTorchBackend` in the `GaussianProcessExact` class due to this). I think it would be ideal to directly inherit if it is possible though an option to workaround might be to create a wrapper (that is also an emulator so it can be used wherever we currently use emulators), e.g.:
```python
class LightningEmulator(L.LightningModule, Emulator):
    def __init__(self, x: TensorLike, y: TensorLike, model: type[Emulator], **kwargs):
        self.model = model(x, y, **kwargs)
```
We doing something like this with `TransformedEmulator` (#474) though might need to revisit how to handle any class/static methods that are model specific.

**Comment by radka-j** [comment]
Thanks for the notes @sgreenbury, that's super useful! I like the look of `lightning-uq-box` too.

**Comment by edwardchalstrey1** [comment]
[fyi](https://chatgpt.com/share/685d511c-8f20-8003-b684-f8e5ece1efe7) you cannot use PyTorch Lightning callbacks directly unless your modules are based on LightningModule and you're using the Trainer, so this isn't something that can be used without doing that, we would need to implement out own callback system

**Comment by sgreenbury** [comment]
Adding notes from discussion with @edwardchalstrey1 on next steps:
- Initial refactor removed `TorchDeviceMixin` - a next step is to check if this can remain and is compatible with `LightningModule`
- For early stopping callback, validation data is needed at fit time. Two potential ways of doing this:
  1. Add validation_data or validation_split vars to `Emulator` API
  2. Perform split (with a seed) within the `_fit()` before the `Trainer`. This relates to the current approach to Tuning where we measure performance with a single fold. This approach would reduce the training data that is currently passed (as it would be being randomly split in fit) although an option might be to pass previously split train/val data.
- Still to look at gpytorch integration

**Comment by edwardchalstrey1** [comment]
Decision for AutoEmulate 1.0 release: park this for now, do #576 instead

**Comment by radka-j** [comment]
No longer plan to use lightning.
